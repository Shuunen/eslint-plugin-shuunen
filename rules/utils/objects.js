/**
 * @param {object | null} object
 * @param {number} maxLevel
 */
function capNesting (object, maxLevel) {
  const one = 1
  /**
   * @param {object | null} currentObject
   * @param {number} currentLevel
   */
  function capHelper (currentObject, currentLevel) {
    if (currentLevel > maxLevel) return "..."
    if (typeof currentObject !== 'object' || currentObject === null) return currentObject
    const result = Array.isArray(currentObject) ? [] : {} // @ts-expect-error
    for (const key in currentObject) if (Object.hasOwn(currentObject, key)) result[key] = capHelper(currentObject[key], currentLevel + one)
    return result
  }
  return capHelper(object, one)
}

module.exports = { capNesting }
