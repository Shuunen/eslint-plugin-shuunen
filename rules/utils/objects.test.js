/* eslint-disable no-magic-numbers */
/* eslint-disable unicorn/no-null */
/* eslint-disable id-length */
import { expect, it } from 'vitest'
import { capNesting } from './objects'

const maxLevel = 3

it('capNesting A - Simple object', () => {
  const deepObject = {
    a: {
      b: {
        c: {
          d: {
            e: {
              f: 'too deep',
            },
          },
        },
      },
    },
  }
  expect(capNesting(deepObject, maxLevel)).toMatchInlineSnapshot(`
    {
      "a": {
        "b": {
          "c": "...",
        },
      },
    }
  `)
})

it('capNesting B - Complex object', () => {
  const deepObject = {
    a: {
      b: {
        c: {
          d: {
            e: {
              f: 'too deep',
            },
          },
        },
        cNull: null,
        cValue: 'value',
      },
      bArray: [1, 2, 3],
      bNull: null,
      bValue: 'value',
    },
    g: {
      h: {
        i: {
          j: {
            k: {
              l: 'too deep',
            },
          },
        },
      },
    },
  }
  expect(capNesting(deepObject, maxLevel)).toMatchSnapshot()
})
