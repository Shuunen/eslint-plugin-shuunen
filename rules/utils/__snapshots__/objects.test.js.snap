// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`capNesting B - Complex object 1`] = `
{
  "a": {
    "b": {
      "c": "...",
      "cNull": "...",
      "cValue": "...",
    },
    "bArray": [
      "...",
      "...",
      "...",
    ],
    "bNull": null,
    "bValue": "value",
  },
  "g": {
    "h": {
      "i": "...",
    },
  },
}
`;
